### PR TITLE
Define non-positive top_k; top_k range check

### PR DESCRIPTION
**Pull request by ivanstepanovftw**
1. With `--top_k 0` expected that all logits will be sampled (by top_p)
2. Range check added to prevent from resizing logits more than vocab have. Resizing to more than vocab have causes token 0 to be appended to logits to be sampled

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -1229,7 +1229,8 @@ static llama_vocab::id llama_sample_top_p_top_k(
         }
     }
 
-    sample_top_k(logits_id, top_k);
+    if (top_k > 0 && top_k < n_logits)
+        sample_top_k(logits_id, top_k);
```
Comment (ggerganov) [inline llama.cpp:1233]:
```suggestion
    if (top_k > 0 && top_k < n_logits) {
        sample_top_k(logits_id, top_k);
    }
```

**Review by ggerganov — APPROVED** [review]

File: llama.cpp (hunk 1 above)
Comment (ivanstepanovftw) [reply llama.cpp]: Sorry, I will match to code style later on

**Comment by Piezoid** [comment]
@ivanstepanovftw @ggerganov 
Correct me if I'm mistaken, but I think the top_p filter needs the logits to be sorted. With this PR, disabling the top_k filter bypasses the sorting.

Instead, the whole array should be sorted when there is no top_k filtering:
```c++
sample_top_k(logits_id, top_k > 0 ? std::min(top_k, n_logits) : n_logits);
```

**Comment by ivanstepanovftw** [comment]
You are right!
